### ISSUE TITLE
ERR: expected to read 4043309056 bytes but read 4294967295

### ISSUE THREAD

**Issue by ssbarnea** [ggreer/the_silver_searcher#1038]
It seems that that ag fails due to malloc with a really weird error when it encounters big files. The issue should be easy to replicate, just assure you have a DVD ISO image in the search path.

**Comment by yahooguntu** [comment]
I'm seeing similar behavior on version 1.0.2 installed via Brew. The issue went away when I moved a 4.4GB file out of the search folder.

**Comment by philip-wernersbach** [comment]
I'm seeing the same thing, and can confirm that this does not happen when I move large files out of the search folder.

```
ag version 1.0.2

Features:
  +jit +lzma +zlib
```

**Comment by AAAton** [comment]
Same issue with ag version 1.0.2 installed from brew. Deliberately trying to search a large file though, so moving it out of the path is out of the question.

**Comment by ssbarnea** [comment]
I would be more than happy to have a default behaviour that is not searching in big files at all. I think that in less than 1% of the cases someone wants for search text in a file bigger than 10MB.

**Comment by AdrienGiboire** [comment]
@ssbarnea I disagree with the 1%. Searching through big files (eg. dump, logs and stuff) is more frequent than you might think.

Plus, the original issue is that `ag` fails when dealing with big files.

**Comment by cwonrails** [comment]
Getting the same error here with `ag` installed via homebrew. `ag -V` results are the same as @philip-wernersbach:

```
ag version 1.0.2

Features:
  +jit +lzma +zlib
```

Here's an example output from my `$HOME` directory:

```
ag -l silver
AUTHORS
ERR: expected to read 2999159722 bytes but read 4294967295
```

Any chance this could be a result of also having `rg` installed?

**Comment by JohnMaguire** [comment]
Also using the `brew` installed version.

```
-> % ag -V
ag version 1.0.2

Features:
  +jit +lzma +zlib
```

Same thing when I ran it against my `Downloads` directory (where I had a bunch of log files I wanted to search). A coworker mentioned that this happens when run against directories with large files. I ended up using `ripgrep` for this situation instead, which had no issues.

**Comment by jbergstroem** [comment]
can reproduce:
```console
$ ag -c HIT access.log 
ERR: expected to read 1401012463 bytes but read 4294967295
$ ag -V               
ag version 1.0.3

Features:
  +jit +lzma +zlib
```

**Comment by diversario** [comment]
Yeah, I installed it via brew to run on multi-gigabyte files and it immediately errored out.
```
$ ag --version
ag version 2.0.0

Features:
  +jit +lzma +zlib
```

**Comment by ssbarnea** [comment]
Should I also add all those errors cause by broken symlinks, missing permissions or files that are sockets? Somehow I find those errors really useless:

```
ERR: Skipping ./.local/lib/python2.7/site-packages/site.py: Error fstat()ing file.
ERR: Skipping ./.local/lib/python2.7/site-packages/sitecustomize.py: Error fstat()ing file.
ERR: Skipping ./.ssh/cp/39798889478c51e2309e4fbfc2afb546dd35368c: Mode 49536 is not a file.
ERR: Skipping ./.ssh/cp/a6692ecdd90b1ba88984394e0605c4ae347e22f0: Mode 49536 is not a file.
ERR: Skipping ./bin/subl: Error fstat()ing file.
ERR: Skipping ./dotfiles/bin/subl: Error fstat()ing file.
ERR: expected to read 4043309056 bytes but read 4294967295
```

**Comment by geoffwa** [comment]
Reading the source this problem only appears to show up whenever `--no-mmap` is set, which is the default on OSX. Running with --mmap 'fixes' the issue. 

Also the [error](https://github.com/ggreer/the_silver_searcher/blob/d10085cbccd602f15b6eca25880b17f5d20ab818/src/search.c#L352)  is a bit misleading - the code looks like: ```c
off_t f_len = 0;
size_t bytes_read = read(fd, buf, f_len);
if ((off_t)bytes_read != f_len) {
    die("expected to read %u bytes but read %u", f_len, bytes_read);
}
```

`%u` implies an int which maxes out at 4 bytes wide, but off_t and size_t (and ssize_t) are all 8 bytes wide on modern OSXen.

**Comment by szha** [comment]
@ggreer

**Comment by minusf** [comment]
this is still an issue.  if what the man page says is still true:

`--[no]mmap` Toggle use of memory-mapped I/O. Defaults to true on platforms where mmap()  is  faster  than  read(). (All but macOS.)

then i think it would be better to be slower than to generate errors... what's the point of speed if it doesn't work?

```
$ ag -V
ag version 2.1.0

Features:
  +jit +lzma +zlib
```

**Comment by jedeleh** [comment]
Running with the -mmap flag fixed it for me, but that's kind of a workaround?

**Comment by Miserlou** [comment]
Still getting this, -mmap cured it but really wish I didn't have to..

``bash
~/Projects/data-refinery $ a RSEM
ERR: expected to read 3403130363 bytes but read 4294967295
~/Projects/data-refinery $ ag -V
ag version 2.1.0

Features:
  +jit +lzma +zlib
~/Projects/data-refinery $ ag RSEM --mmap
workers/data_refinery_workers/processors/transcriptome_index.py
155:    # RSEM takes a prefix path and then all files generated by it will
247:      * Next the tool RSEM's prepare-reference is run.

workers/Dockerfile.base
48:RUN git clone https://github.com/deweylab/RSEM.git
50:RUN cd RSEM && make install
52:RUN rm -rf RSEM
```

I don't care that `ag` couldn't scan the huge file, but I _do_ mind that because of that one file, I didn't get any results at all. Handle the exception better and scan the rest!

**Comment by danshumaker** [comment]
Throwing in 2 cents here.  The MAIN reason I use ag is for speed searching large amounts of files and large files.  So I'd say the use case is more like 100%.  Otherwise, we'd all just back to using grep.  P.s. I absolutely love the silver searcher!

**Comment by danemacmillan** [comment]
I'll just mention that there is a tool called ripgrep (`rg`), and believe it or not, is actually faster than `ag`. It also has no issue looking through massive DB dumps, for example, which is why many of us are here reporting about this shortcoming. I had been using `ag` for years until I found `rg`. You guys should give it a try. I haven't looked back.

https://github.com/BurntSushi/ripgrep

**Comment by calvinjtaylor** [comment]
I experienced the subject error using v2.1, and quickly found this thread and updated to v2.2.  The error has gone away, but now the search never completes in a timely maner like it used to.  I'm not sure what changed, I'm searching a really large git repository, and there is a 2G 
find . -size +500000 -print -exec ls -lh {} \;
./.git/objects/pack/pack-c502cb0b2fcb5505e8ca0c1cdc3f77e134aa8eba.pack
-r--r--r--  1 me@here  heretoo   2.0G  6 Nov 12:23 ./.git/objects/pack/pack-c502cb0b2fcb5505e8ca0c1cdc3f77e134aa8eba.pack

**Comment by intel352** [comment]
On OSX, agreed v2.2 fixed this error, but now running a search seems much slower

**Comment by intel352** [comment]
@danemacmillan Thanks for mentioning `rg`, I'm switching over to that from `ag`. As of v2.2, `ag`'s default performance on OSX is *ridiculously* slow, to the point I feel it's getting hung. I have to run it with `--mmap` flag to get it to return in an acceptable amount of time, whereas `rg` actually works.

Thanks!
